internal/controller/shim_controller.go

```diff
@@ -136,6 +138,11 @@ func (sr *ShimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 
 	// 4. Deploy job to each node in list
 	if len(nodes.Items) > 0 {
+		// TODO(flavio): the handleInstallShim returns an error as soon as the
+		// creationg of one of the CronJob fails. We create on CronJob per node,
+		// hence failing on node 5/10 will cause this reconciliation loop to
+		// exit immediately. However, the shim finalizer has not been set yet,
+		// that happens at end of this `if` block.
```

Comment:
I would propose to change the `handleInstallShim` to return an array of errors, so that all the nodes get the chance to get the shim installed on them.

Then OFC we need to ensure the finalizer is always set, even if there was an error

Reply:
I agree that calling the installation deployJobOnNode should happen concurrently, and potential errors should be returned aggregated to help implement potential strategies, e.g., retries later on if needed.

Also, it would indeed make sense to call ensureFinalizerForShim even if an error happens:

`
    // Ensure the finalizer is called even if a return happens before
    defer func() {
        err := sr.ensureFinalizerForShim(ctx, &shimResource, KwasmOperatorFinalizer)
        if err != nil {
            log.Error().Msgf("Failed to ensure finalizer: %s", err)
        }
    }()`